### PR TITLE
DTSRD-1879: Case flags 2.1 ITHC Observation "Broken Access Control"

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/DTSRD-1879

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Review by SabinaHMCTS — APPROVED** [review]

**Review by lukasz-wolski — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImpl.java [hunk 1]
```diff
@@ -66,6 +74,24 @@ public CaseFlag retrieveCaseFlagByServiceId(String serviceId, String flagType,
...
+        }
+        return StringUtils.isEmpty(availableExternalFlag)
+            || availableExternalFlag.trim().equalsIgnoreCase("y");
+    }
+
+    private List<CaseFlagDto> filterCaseFlags(List<CaseFlagDto> caseFlagDtoList, boolean isAvailableExternalFlag) {
+        if (!isAvailableExternalFlag) {
+            return caseFlagDtoList;
+        }
+        return caseFlagDtoList.parallelStream().filter(caseFlagDto -> {
```
[the remark this answers is not in the document]
Comment (lukasz-wolski) [reply src/main/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImpl.java:87]: see https://stackoverflow.com/a/20375622/2754462 for a discussion of considerations when to use parallel streams, 3rd point: 'if I already have many requests to process in parallel, adding an additional layer of parallelism inside each request could have more negative than positive effects)'

File: src/main/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImpl.java [hunk 2]
```diff
@@ -66,6 +74,24 @@ public CaseFlag retrieveCaseFlagByServiceId(String serviceId, String flagType,
...
+            Boolean externallyAvailable = caseFlagDto.getExternallyAvailable();
+            return externallyAvailable == null || externallyAvailable.booleanValue() == true;
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImpl.java:89]: done

File: src/test/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImplTest.java [hunk 3]
```diff
@@ -177,6 +197,95 @@ void testGetCaseFlag_ByServiceIWithWelshRequiredIsNandAvailableExternallyIsN_Ret
...
+                                                                   welshRequired,
+                                                                   availableExternalFlag);
+        validateCaseFlags(caseFlag, "N");
+    }
+
+    private void validateCaseFlags(CaseFlag caseFlags, String flag) {
+        boolean externallyAvailable = (StringUtils.isNotEmpty(flag) && (flag.trim().equalsIgnoreCase("y")));
+        assertNotNull(caseFlags);
+        assertNotNull(caseFlags.getFlags());
+        if (externallyAvailable) {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImplTest.java:250]: refactored now

File: src/main/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImpl.java [hunk 4]
```diff
@@ -66,6 +74,24 @@ public CaseFlag retrieveCaseFlagByServiceId(String serviceId, String flagType,
         return caseFlag;
     }
 
+    private boolean availableExternally(String availableExternalFlag) {
+        if (UserInfoUtil.hasPrdRoles(jwtGrantedAuthoritiesConverter.getUserInfo())) {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/service/impl/CaseFlagServiceImpl.java:76]: done

**Review by lukasz-wolski — APPROVED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveCaseFlagsIntegrationTest.java [hunk 5]
```diff
@@ -192,27 +193,17 @@ void shouldReturnSuccessForRetrieveCaseFlagsByServiceIdWithWelshFlagIsY() throws
...
+        Exception exception = assertThrows(UnrecognizedPropertyException.class, () -> {
+            commonDataApiClient.retrieveCaseFlagsByServiceId(
+                "AAA1?flag-type=party&available-external-flag=Y",
+                CaseFlag.class,
+                path
+            );
+        });
+        assertNotNull(exception);
+        assertTrue(exception.getLocalizedMessage().contains("Data not found"));
     }
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveCaseFlagsIntegrationTest.java:208]: i dont see test cases for prd roles scenario and and also there are no assertions for available-external-flag requests for parent flag and all child flags.

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/cdapi/oidc/JwtGrantedAuthoritiesConverter.java [hunk 6]
```diff
@@ -56,4 +58,8 @@ private List<GrantedAuthority> extractAuthorityFromClaims(List<String> roles) {
             .map(GrantedAuthority.class::cast)
             .toList();
     }
+
+    public UserInfo getUserInfo() {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/oidc/JwtGrantedAuthoritiesConverter.java:62]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveCaseFlagsIntegrationTest.java (hunk 5 above)
Comment (kiran-yenigala-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveCaseFlagsIntegrationTest.java]: added integration tests

File: src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveExternallyAvailableCaseFlagsIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    private void validateCaseFlags(CaseFlag caseFlags, boolean externallyAvailable) {
+        assertNotNull(caseFlags);
+        assertNotNull(caseFlags.getFlags());
+        caseFlags.getFlags().stream().forEach(caseFlag -> {
+            assertNotNull(caseFlag.getFlagDetails());
+            List<FlagDetail> flagDetailsList = caseFlag.getFlagDetails();
+            flagDetailsList.stream().forEach(flagDetail -> {
+                boolean flagExternallyAvailable = flagDetail.getExternallyAvailable();
+                if (externallyAvailable) {
+                    MatcherAssert.assertThat(flagExternallyAvailable, anyOf(is(true)));
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveExternallyAvailableCaseFlagsIntegrationTest.java:79]: change it to allOf

File: src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveExternallyAvailableCaseFlagsIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        });
+    }
+
+    private void validateChildFlags(List<FlagDetail> flagDetails,
+                                    boolean externallyAvailable) {
+        if (flagDetails != null) {
+            flagDetails.stream()
+                .forEach(flagDetail -> {
+                    if (externallyAvailable) {
+                        MatcherAssert.assertThat(flagDetail.getExternallyAvailable(), anyOf(is(true)));
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/cdapi/RetrieveExternallyAvailableCaseFlagsIntegrationTest.java:95]: change it to allOf

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

**Review by SabinaHMCTS — APPROVED** [review]

**Review by SabinaHMCTS — APPROVED** [review]

**Review by rajkatla-hmcts — APPROVED** [review]

**Pull request merged**